Fix OverflowException in JsonSerializer.Serialize

felipepessoto opened the pull request:
Fix https://github.com/dotnet/runtime/issues/609

felipepessoto committed:
Add failing test

At first I tried to create five internal references to CustomClassToExceedMaxBufferSize and create 25 levels deep instances, but the tests ran forever. So I changed it to have several big string properties, which lead to the System.OverflowException using 3 level deep instance.

felipepessoto committed:
Initial implementation. Still needs to throw the appropriated Exception

felipepessoto committed:
Validate new buffer size against MaxByteArrayLength

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs:
```diff
@@ -138,7 +139,12 @@ private void CheckAndResizeBuffer(int sizeHint)
+                if ((uint)newSize > MaxByteArrayLength)
+                {
+                    throw new JsonException();
+                }
```
This is throwing if _rentedBuffer.Length + growBy is too large, but that doesn't necessarily mean that _rentedBuffer.Length + sizeHint is too large.

It also seems like this should be an OutOfMemoryException, rather than a JsonException.  That's what the runtime would throw if too large a value were passed when new'ing the array, and this case really isn't different.

Consider using logic closer to that of `List<T>.EnsureCapacity`.
https://github.com/dotnet/runtime/blob/4d3ec15a0e197836f46d4297a5c6ae9722ec26f7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L405-L409

felipepessoto replied:
@stephentoub List doesn't handle this specific scenario. It only validates the "auto-grow" length, but not "min" parameter.

If you run

    var a = new List<int>();
    a.Capacity = int.MaxValue;

It tries to allocate the array and throws an OutOfMemoryException.

But we are trying to throw a more user-friendly Exception, so I can validate the size and throw an OutOfMemoryException with a custom message. Something like:

    int growBy = Math.Max(sizeHint, _rentedBuffer.Length);

    int newSize = _rentedBuffer.Length + growBy;

    if ((uint)newSize > MaxByteArrayLength && growBy > sizeHint)
    {
        newSize = _rentedBuffer.Length + sizeHint;
    }

    if ((uint)newSize > MaxByteArrayLength)
    {
        //newSize = MaxByteArrayLength;
        ThrowHelper.ThrowJsonException_ResultExceededMaximumBufferSize();//OutOfMemory
    }

I have a question, GetSpan and GetMemory can return a buffer smaller than sizeHint? Because CheckAndResizeBuffer has a Assert:

    Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);

So I can't set newSize to MaxByteArrayLength. But some other classes validates the buffer returned by GetMemory: https://github.com/dotnet/runtime/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L1082

stephentoub replied:
> But we are trying to throw a more user-friendly Exception

Why isn't an OutOfMemoryException sufficient?

felipepessoto replied:
You are right, OutOfMemory should be obvious enough

ahsonkhan replied:
> I have a question, GetSpan and GetMemory can return a buffer smaller than sizeHint? Because CheckAndResizeBuffer has a Assert:

No, GetSpan/GetMemory shouldn't return anything smaller than sizeHint.
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1.getspan?view=netcore-3.1#System_Buffers_IBufferWriter_1_GetSpan_System_Int32_

The `Utf8JsonWriter` is doing extra checks for it, since someone could implement an `IBufferWriter` incorrectly that doesn't honor the contract and pass that to the `public` API. In this case, `PooledBufferWriter` is internal and we ensure it honors the contract or throws.

felipepessoto committed:
When newSize overflows because growBy == _renteBuffer.Length, it tries to use growBy = sizeHint
If it still overflow, we set the newSize to int.MaxValue, which will throw an OutOfMemoryException when calling ArrayPool.Rent.

felipepessoto commented:
/azp run

danmosemsft commented:
/azp run

stephentoub commented:
@ahsonkhan, @steveharter, can you take a look as well?  Thanks.

ahsonkhan commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs:
```diff
@@ -138,7 +138,17 @@ private void CheckAndResizeBuffer(int sizeHint)
-                int newSize = checked(_rentedBuffer.Length + growBy);
+                int newSize = _rentedBuffer.Length + growBy;
+
+                if ((uint)newSize > int.MaxValue && growBy > sizeHint)
```
nit: Consider re-writing it as follows instead (smaller assembly and reduces the extra checks in the common case):
- Use a local for the length
- Nest the second check inside the first.
- Remove the `growBy > sizeHint` condition, and optimistically assume  that if `newSize` is too large its because we grew by `_rentedBuffer` and not by `sizeHint` and that growing by `sizeHint` might succeed.
```C#
int currentLength = _rentedBuffer.Length;
int growBy = Math.Max(sizeHint, currentLength);

int newSize = currentLength + growBy;

if ((uint)newSize > int.MaxValue)
{
    newSize = currentLength + sizeHint;
    if ((uint)newSize > int.MaxValue)
    {
        newSize = int.MaxValue;
    }
}
```

felipepessoto replied:
Nice, I did the 2nd and 3rd suggestion some time ago and undid because was not sure if it was better.

felipepessoto replied:
> Are all these new branches covered by tests? Can you check test coverage by running the following and seeing the coverage report generated (https://github.com/dotnet/runtime/blob/9900dfb4b2e32cf02ca846adaf11e93211629ede/docs/workflow/building/libraries/code-coverage.md#local-code-coverage-runs):
> 
> ```
> dotnet msbuild /t:BuildAndTest /p:Coverage=true
> ```

The tests seems to take forever to start when I enable Coverage, is it normal?

stephentoub replied:
> The tests seems to take forever to start and run when I enable Coverage, is it normal?

They should still start quickly, e.g. you should see the Discovering/Discovered/Starting lines quickly, but the tests can take a long time to run.  Here's what I get with what's in master:
```
    Discovering: System.Text.Json.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Text.Json.Tests (found 1649 of 1680 test cases)
    Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 12)
    Finished:    System.Text.Json.Tests
  === TEST EXECUTION SUMMARY ===
     System.Text.Json.Tests  Total: 11344, Errors: 0, Failed: 0, Skipped: 0, Time: 58.468s
```

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -68,5 +68,37 @@ public static void SerializeToWriterRoundTripEscaping()
+
+            static string name = new string('a', 53687091);
+            public string GetName1 { get { return name; } }
+            public string GetName2 { get { return name; } }
```
nit: Use expression bodied members for all these properties

```suggestion
            public string GetName2 => s_name;
```

ahsonkhan commented:
> Fix #609

We should consider making a similar fix to the public `ArrayBufferWriter` as well, based on where we land in this PR.

https://github.com/dotnet/runtime/blob/fd7a5e8cf1748b7aefa90dd8e6966af2cbb4fe8b/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L168-L180

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs:
```diff
@@ -141,9 +140,14 @@ private void CheckAndResizeBuffer(int sizeHint)
+                    newSize = _rentedBuffer.Length + sizeHint;
+                }
+
+                if ((uint)newSize > int.MaxValue)
```
https://source.dot.net/#System.Private.CoreLib/OutOfMemoryException.cs,11bcfd9ad4159d54,references

felipepessoto replied:
So should I throw the Exception explicitly?

stephentoub replied:
I think it's fine as is.

felipepessoto commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -68,5 +68,37 @@ public static void SerializeToWriterRoundTripEscaping()
+        }
+
+        [Fact]
+        public static void SerializeExceedMaximumBufferSize()
```
It consumes around 4GB. When I run it using the command line, the result says 74 seconds, but Idk if it counts the compilation and discovery time:

dotnet msbuild /t:BuildAndTest /p:XunitMethodName=System.Text.Json.Serialization.Tests.WriteValueTests.SerializeExceedMaximumBufferSize

System.Text.Json.Tests  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 74.272s

stephentoub replied:
At a minimum this should be restricted to 64-bit processes; trying to allocate 4GB in a 32-bit process is not going to go well.

Even so, 4GB is a lot, even for outerloop.

felipepessoto committed:
Change to make assembly smaller and reduce a extra check in the common case

felipepessoto committed:
Name conventions
OuterLoop
Make the test simpler

felipepessoto commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -68,5 +68,37 @@ public static void SerializeToWriterRoundTripEscaping()
+            public CustomClassToExceedMaxBufferSize Internal1 { get; set; }
+            public CustomClassToExceedMaxBufferSize Internal2 { get; set; }
+
+            static string name = new string('a', 53687091);
```
Done

felipepessoto commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -68,5 +68,37 @@ public static void SerializeToWriterRoundTripEscaping()
+        {
+            CustomClassToExceedMaxBufferSize temp = new CustomClassToExceedMaxBufferSize();
+
+            for (int i = 0; i < 2; i++)
```
Done

stephentoub commented:
@felipepessoto, sorry for the delay. This looks good and ready to go, but there's a conflict in the tests now... can you rebase to address it?

felipepessoto commented:
> @felipepessoto, sorry for the delay. This looks good and ready to go, but there's a conflict in the tests now... can you rebase to address it?

Yes sure. But I still have some pending things to do. I still didn't stop to check how can I restrict the test to 64bits, and I couldn't run the coverage test yet, last time it took hours and didn't finish

stephentoub commented:
> I still didn't stop to check how can I restrict the test to 64bits

e.g.
https://github.com/dotnet/runtime/blob/0f834db1fd80cf82e5ef27f72c48af1c911775da/src/libraries/System.Text.Json/tests/Utf8JsonWriterTests.cs#L723
https://github.com/dotnet/runtime/blob/0f834db1fd80cf82e5ef27f72c48af1c911775da/src/libraries/System.Text.Json/tests/Utf8JsonWriterTests.cs#L27

ahsonkhan commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs:
```diff
@@ -132,13 +132,23 @@ private void CheckAndResizeBuffer(int sizeHint)
 
-            int availableSpace = _rentedBuffer.Length - _index;
+            int currentLength = _rentedBuffer.Length;
+            int availableSpace = currentLength - _index;
```
nit: Leave this as is and move the `int currentLength` local into the `if (sizeHint > availableSpace)` check.

```suggestion
            int availableSpace = _rentedBuffer.Length - _index;
```

[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5ilwDcNGsQDMTUgwDCDAN40GihgqUAHKAEsAbtgwwGwAJ56A2gF0GAfVgA7PQBNOPPkOpKG67bv0a7V3/YwCK7uKoqeOnpMKNIAFjBgANYAgjb2AEowuBoAXjBOvFAAFL4YDNl5ABKlAJRhcvXuGlwMRRUw1X4AvF0MdHVu7kryg0ND7Z1lvaQArEghYwwAvsKjQ6UMYBxQthgAMjA2AOYYsQy91ocOBXwsB8enC2Mb2DoaADbYwO8wAMqq2DA+l6Wx2V3uJzOcH8aSCrkaSmarQmGwAfAxXtgPl8fv9ATABosGmsiRsjlAIAB3NiGc4MACyuliLEZCDauQ6pTQm22uwhpxq8JJiw2NhglN+HLpoL5h0hDAA1AxyVSaUKiU0WkUihxamKJVL0aUWdgEAA1bDvDgEhFjEYaxb6yV5aW88Fy06K8ocyZPB2IrU6vXi536I12E3my3Wwn+9z2uOOkNS3rG1kWq0wP1xlbCsa521DIymCwQd6ObiFOmXOwwCvOKDq/0166Vvh05I7bCGAAKEDLAB5izBUSxfrFsLB7CxMnYik6OYLVnG8TYh8YRx5YFoNBAOLgblA6WX64UWMlcKuinRuZYAkEl3mhuoYDu9we21AWFIIKpDAAVCAihbOtD0fOMXzffdD2/H5JyKcD/U7KBuz7Qdh1HcdJzrGcYAwbYbCKE8wOzJRc3ccilHqCJvGiOIEhSNJMnaQ9LFwDgjiOLIHBKPwUTsWNFATZ4tX4qZen6QthI1MS6VmeZCwLJ8XjeT5vj+AEgWrXZQM/O4PShGFAmCZcRVEn00QxVScQ0/FBKGaTST8GV3QeM4Lh009bn5WJSOePwVWpWlekZU5I3ZKouR5ME7B8x9CyaPwFxdEE3VigyvUCtVTIdJFtV1ATkrDBg01NDMYwS4ZKrGIrXRi/YMqVMS/I1PKg0K5MXXDDBI3Km0nyJRzE0UWrUwjdNoyzaqhkoh0lLjYdzAYYjP20q5dIbJsHRAryj16ZDUP7d51z0TCJynXC5yK+KBqUVcTs3SDd2g1behWhtz0vAFCJvIyHxa9wnvfGCfz/QDgM8kjpqBl6PqkODikQh0Dt7I6HrO7Dp0yfCoEI97CiRmb6lzJYgA=)

ahsonkhan commented:
> I couldn't run the coverage test yet, last time it took hours and didn't finish

You are probably building using a debug coreclr which causes tests to take a long time to run. If you build coreclr for release and then build the libraries using the release coreclr, the tests should run faster.

**I was able to run the coverage after merging master into your branch:**
From root of repo run the following:
`build.cmd -subsetCategory coreclr -c Release && build.cmd -subsetCategory libraries /p:CoreCLRConfiguration=Release`
Then navigate to `src\libraries\System.Text.Json\tests` and run:
`dotnet msbuild /t:BuildAndTest /p:Coverage=true`

See https://github.com/dotnet/runtime/issues/839#issuecomment-565684651

cc @safern, @ViktorHofer - do we need to increase the visibility of this issue so fewer folks hit against running tests with debug coreclr, by default?

Looks like all the new branches introduced in this PR are covered by the test :)
![image](https://user-images.githubusercontent.com/6527137/74572152-861a0500-4f33-11ea-91f8-3b979aaad0f1.png)

The path that is likely not covered (and this sequence of code flow won't show up in test coverage):
`(uint)(currentLength + growBy) > int.MaxValue` is **true** but `(uint)(currentLength + sizeHint) > int.MaxValue` is **false**

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -68,5 +68,38 @@ public static void SerializeToWriterRoundTripEscaping()
+            public string GetName17 => s_name;
+            public string GetName18 => s_name;
+            public string GetName19 => s_name;
+            public string GetName20 => s_name;
```
You don't need to use up 4 GB to test the scenario. Anything over ~2GB would be sufficient.

Keeping 11 copies of the string would work as well (you can remove the `GetName12` to `GetName20`).

felipepessoto replied:
If I reduce it, this code is not executed:

![image](https://user-images.githubusercontent.com/1336227/74621179-ca9ecf80-50f0-11ea-83d8-015b8fa4e3b1.png)

felipepessoto replied:
> The path that is likely not covered (and this sequence of code flow won't show up in test coverage):
> `(uint)(currentLength + growBy) > int.MaxValue` is **true** but `(uint)(currentLength + sizeHint) > int.MaxValue` is **false**

We can cover that path, using the class with less than 20 name fields, but I'm not sure if we need it.

ahsonkhan replied:
> but I'm not sure if we need it.

It wouldn't hurt to add a test for that as well. It does improve our "path" coverage, in case the grow logic were to change.

felipepessoto replied:
Actually, with 20 Fields, both branches are covered:

![image](https://user-images.githubusercontent.com/1336227/74903384-9bd96100-535d-11ea-8c12-600d07d9bd05.png)

ahsonkhan replied:
Nice.

felipepessoto committed:
Improve assembly size
Run test on 64 bits only

felipepessoto commented:
> > I still didn't stop to check how can I restrict the test to 64bits
> 
> e.g.
> https://github.com/dotnet/runtime/blob/0f834db1fd80cf82e5ef27f72c48af1c911775da/src/libraries/System.Text.Json/tests/Utf8JsonWriterTests.cs#L723
> 
> 
> https://github.com/dotnet/runtime/blob/0f834db1fd80cf82e5ef27f72c48af1c911775da/src/libraries/System.Text.Json/tests/Utf8JsonWriterTests.cs#L27

Thx, done

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs:
```diff
@@ -285,5 +287,38 @@ static void WriteAndValidate(object input, Type type, string expected, JsonSeria
+            public string GetName20 => s_name;
+        }
+
+        [ConditionalFact(nameof(IsX64)), OuterLoop]
```
Skip running this test on linux.

See other examples (feel free to copy the comment):
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.ValueTextEquals.cs#L631-L638

felipepessoto committed:
Added Platform Specific attribute.
Separate OuterLoop attribute, to be consistent with the other tests in this solution.

ahsonkhan commented:
@felipepessoto - are you up for making a similar fix to `ArrayBufferWriter<T>`, in a separate PR to fix/close the issue?

https://github.com/dotnet/runtime/blob/4f7108147c25e8e89dc16864d7f439e6cba7470b/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L158-L183

felipepessoto commented:
> @felipepessoto - are you up for making a similar fix to `ArrayBufferWriter<T>`, in a separate PR to fix/close the issue?
> 
> https://github.com/dotnet/runtime/blob/4f7108147c25e8e89dc16864d7f439e6cba7470b/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L158-L183

Absolutely
